src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -89,18 +90,40 @@ public function __construct(ParameterBagInterface $parameterBag = null)
      */
     public function compile()
     {
+        if ($this->isCompiled()) {
+            return;
+        }
+
         $this->parameterBag->resolve();
 
-        $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+        if (!$this->parameterBag instanceof FrozenParameterBag) {
+            $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+        }
+
+        $this->compiled = true;
+    }
+
+    /**
+     * Returns true if the container is compiled.
+     *
+     * @return bool
+     */
+    public function isCompiled()
+    {
+        return $this->compiled;
     }
 
     /**
      * Returns true if the container parameter bag are frozen.
      *
+     * Deprecated since 3.2, to be removed in 4.0.
+     *
      * @return bool true if the container parameter bag are frozen, false otherwise
      */
     public function isFrozen()
     {
+        @trigger_error(sprintf('The %s method is deprecated as of 3.2 and will be removed in 4.0. Please use "%s" instead.', __METHOD__, __CLASS__.'::isCompiled'), E_USER_DEPRECATED);
```

Comment:
`The %s() method is deprecated as of 3.2 and will be removed in 4.0. Use the isCompiled() method instead.`
Lets save repeating the (looong) namespace. And we don't "please" usually here :)